scripting/include/srccoop/manager.inc

```diff
@@ -170,123 +234,47 @@ methodmap CoopManager
 			else if (g_pLevelLump.m_iIntroType == INTRO_FREEZE)
 			{
 				pPlayer.SetFlags(pPlayer.GetFlags() | FL_ATCONTROLS);
-				Client_ScreenFade(pPlayer.GetEntIndex(), Conf.FROZEN_FADE_DUR_IN, FFADE_OUT|FFADE_STAYOUT, _,
+				Client_ScreenFade(pPlayer.GetEntIndex(), Conf.FROZEN_FADE_DUR_IN, FFADE_OUT | FFADE_STAYOUT, _,
 					Conf.FROZEN_FADE_COLOR[0], Conf.FROZEN_FADE_COLOR[1], Conf.FROZEN_FADE_COLOR[2], Conf.FROZEN_FADE_COLOR[3]);
 			}
-			CoopManager.CheckStart();
-		}
-	}
 
-	public static void CheckStart()
-	{
-		if (!data.m_bStarted && data.m_bPlayerSpawned)
-		{
-			// are we allowed to start or waiting for players?
-			if (CoopManager.IsAllowedToStart())
+			if ((g_pLevelLump.m_iIntroType == INTRO_NONE && !data.m_bDoesMapHaveDelayedOutput))
 			{
-				// is initial timer already running?
-				if (CoopManager.IsInCountDownPeriod())
-				{
-					// can start, but do a 3 sec readyup timer as we cancel the initial player wait timer
-					if (!data.m_bGetReadyTimer)
-					{
-						delete data.m_pStartTimer;
-						CoopManager.StartCountdown(true);
-					}
-				}
-				else
+				delete data.m_hStartTimer;
+				data.m_iSecondsToStart = 0;
+				Timer_DeincrementUntilMapStart(null);
+			}
+			else if (((GetGameTime() - data.m_flMapStartTime) <= g_flWaitPeriodSkipGraceTime) && (data.m_iSpawnedPlayerCountLastMap - data.m_iSpawnedPlayerCount) == 0)
+			{
+				if (!data.m_hStartTimer || data.m_iSecondsToStart > 3)
 				{
-					// we can just start if no previous timer is running
-					CoopManager.Start();
+					delete data.m_hStartTimer;
+					data.m_iSecondsToStart = 3;
+					CoopManager.SetIntroTimeHud(data.m_iSecondsToStart);
+					data.m_hStartTimer = CreateTimer(1.0, Timer_DeincrementUntilMapStart, _, TIMER_FLAG_NO_MAPCHANGE | TIMER_REPEAT);
 				}
 			}
 			else if (!CoopManager.IsInCountDownPeriod())
 			{
-				// not allowed to start now and no timer is running, start the timer.
-				CoopManager.StartCountdown(false);
+				data.m_iSecondsToStart = g_pConvarWaitPeriod.IntValue;
+				CoopManager.SetIntroTimeHud(data.m_iSecondsToStart);
+				data.m_hStartTimer = CreateTimer(1.0, Timer_DeincrementUntilMapStart, _, TIMER_FLAG_NO_MAPCHANGE | TIMER_REPEAT);
```

Comment:
It's not a bad thing to split functions to indicate intention, e.g. StartIntroTimer. These are not performance critical.

Reply:
Created a new function `StartIntroTimer` which encapsulates the creation of the HUD timer and countdown.